DOCSP-40066: Support SelectMany inside Project/Select (#373)
mongodb/docs-csharp: source/fundamentals/linq.txt

   { "date" : ISODate("2012-12-05T00:00:00Z"), "grade" : "A", "score" : 13 }
   { "date" : ISODate("2012-05-17T00:00:00Z"), "grade" : "A", "score" : 11 }

Nested Statements
+++++++++++++++++

You can chain or nest ``Select`` and ``SelectMany`` statements. The examples

> in this section use the ``Guidebook`` class, which contain an array of ``Restaurant``

Reviewer: S:
```suggestion
in this section use the ``Guidebook`` class, which contains an array of ``Restaurant``
```

objects. The following code shows an example of a document modeled by the 

> ``Guidebook`` class:

Reviewer: Q: do you need to use another class? I think it makes this section a little long compared to the others; if you do need Guidebook, you could move it to "Overview"

.. code-block:: json

   { 
      "_id": { "$oid": ... },
      "restaurants": [
         {
            "_id": { ... } ,
            "address": { ... },
            "name": "Tov Kosher Kitchen",
            ...
         },
         {
            "_id": { ... } ,
            "address": { ... },
            "name": "Harriet's Kitchen",
            ...
         },
         ...
      ]
   }

The following code models the ``Guidebook`` class:

.. literalinclude:: /includes/fundamentals/code-examples/linq.cs
   :language: csharp
   :start-after: start-guidebook-model
   :end-before: end-guidebook-model

You can nest ``SelectMany`` statements within ``SelectMany`` or ``Select``

> statements. The following example nests a ``SelectMany`` statement within a

Reviewer: Q: can you briefly explain why you'd want to nest these / what nesting the statements does? From my understanding (I'm not super familiar with c#), it would be something like:
```suggestion
statements to process nested document fields. The following example nests a ``SelectMany`` statement within a
```

``Select`` statement to retrieve an array of grades from all restaurants
in a document modeled by the ``Guidebook`` class.

> 

Reviewer: I: since the examples on this page use the "sample_restaurants.restaurants" collection, which the "queryableCollection" variable represents, I think it gets confusing to introduce "Guidebooks". "Restaurants" kinda has to be the top-level class, since each instance represents a "restaurants" document. Is there a way to nest SelectMany within Select without introducing Guidebooks -- maybe selecting for restaurant documents with a certain "cuisine" value, then unwinding the grades arrays for just those documents?

Author: While I agree that it is confusing to introduce "Guidebooks," unfortunately filtering by value requires `Where` instead of `Select`. The ideal example for this would involve nested arrays within a document. I tried to update the example to be a bit less involved, but I'm not sure if it's better or worse.

Reviewer: ah okay! Then I think it's fine to introduce a new sample document

This field contains an array of documents that represent the ``Restaurants`` class:

.. code-block:: json

   { 
      "_id": { "$oid": ... },
      "restaurants": [
         {
            "_id": { ... } ,
            "address": { ... },
            "name": "Tov Kosher Kitchen",
            "grades": [
               { 
                  "date" : ISODate("2014-11-24T00:00:00Z"), 
                  "grade" : "Z", 
                  "score" : 20.0 
               }, 
               { 
                  "date" : ISODate("2013-01-17T00:00:00Z"), 
                  "grade" : "A", 
                  "score" : 13.0 
               },
               ...
            ]
            ...
         },
         {
            "_id": { ... } ,
            "address": { ... },
            "name": "Harriet's Kitchen",
            "grades": [ ... ],
            ...
         },
         ...
      ]
   }

You can nest ``SelectMany`` statements within ``SelectMany`` or ``Select``
statements. The following example nests a ``SelectMany`` statement within a
``Select`` statement to retrieve an array of ``Grade`` documents from all restaurants

> in the document above.

Reviewer: S: "preceding" instead of "[above](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#above)":
```suggestion
in the preceding document.
```

   { "date" : ISODate("2012-12-05T00:00:00Z"), "grade" : "A", "score" : 13 }
   { "date" : ISODate("2012-05-17T00:00:00Z"), "grade" : "A", "score" : 11 }

Nested Statements
+++++++++++++++++

You can chain or nest ``Select`` and ``SelectMany`` statements to unwind nested
arrays. Consider a collection that contains documents with a **new** schema. These
documents contain a ``restaurants`` field, which holds an array of documents 

> that each represent the ``Restaurant`` class. The documents within the array each have

Reviewer: S: since the classes model the documents, I think it's more accurate to say the document is "represented by" Restaurants, rather than represent Restaurants
```suggestion
represented by the ``Restaurant`` class. The documents within the array each have
```

> a ``grades`` field that holds an array of documents that each represent
> the ``Grade`` class. The following code is an example of a single document in

Reviewer: S: same suggestion as above
```suggestion
a ``grades`` field that holds an array of documents represented by
the ``Grade`` class. The following code is an example of a single document in
```

            "grades": [
               { 
                  "date" : ISODate("2014-11-24T00:00:00Z"), 
                  "grade" : "Z", 
                  "score" : 20.0 
               }, 
               { 
                  "date" : ISODate("2013-01-17T00:00:00Z"), 
                  "grade" : "A", 
                  "score" : 13.0 
               },
               ...
            ]
            ...
         },
         {
            "_id": { ... } ,
            "address": { ... },
            "name": "Harriet's Kitchen",
            "grades": [ ... ],
            ...
         },
         ...
      ]
   }

You can nest ``SelectMany`` statements within ``SelectMany`` or ``Select``
statements. The following example nests a ``SelectMany`` statement within a
``Select`` statement to retrieve an array from each document in the collection.
Each array holds all grade objects from all restaurants in each document.

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-examples/linq.cs
      :language: csharp
      :start-after: start-nested-SelectMany
      :end-before: end-nested-SelectMany

   .. output::
      :visible: false
      :language: json

>       // output for first document in collection

Reviewer: Q: should this be the first restaurant instead? since these are just the grades for Tov Kosher Kitchen, but the first guidebook document would have more restaurants
```suggestion
      // output for first restaurant
```

Author: Nope, it should be first document - I will make the example/output more clear
